### PR TITLE
fixed schema for table; fixed _concatenate_tables()

### PR DESCRIPTION
**Pull request by LucaMarconato**
Addressed this comment: https://github.com/scverse/spatialdata/issues/158#issuecomment-1447223143.

Ready for review and being merged.

### CODE REVIEW

**Comment by giovp** [comment]
@LucaMarconato I think the logic that you outlined in the comment is correct, I confirmed it here

https://github.com/scverse/spatialdata/issues/158#issuecomment-1447767542

File: spatialdata/_core/models.py [hunk 1]
```diff
@@ -629,7 +629,6 @@ def validate(
                 if "instance_key" not in attr:
                     raise ValueError("`instance_key` not found in `adata.uns['spatialdata_attr']`.")
             elif isinstance(attr["region"], str):
-                assert attr["region_key"] is None
```
Comment (giovp) [inline spatialdata/_core/models.py:632]: this should not be removed, it's correct

File: spatialdata/_core/models.py [hunk 2]
```diff
@@ -656,10 +655,6 @@ def parse(
             instance_key = attr[cls.INSTANCE_KEY]
 
         if isinstance(region, str):
-            if region_key is not None:
```
Comment (giovp) [inline spatialdata/_core/models.py:659]: this should not be removed, it's correct

File: tests/conftest.py [hunk 3]
```diff
@@ -251,14 +251,16 @@ def _get_table(
...
-        region_key = region_key or "annotated_region"
         instance_key = instance_key or "instance_id"
     adata = AnnData(RNG.normal(size=(100, 10)), obs=pd.DataFrame(RNG.normal(size=(100, 3)), columns=["a", "b", "c"]))
     if instance_key is not None:
+        # from today pycharm refuses to execute this line if I run a test...
         adata.obs[instance_key] = np.arange(adata.n_obs)
+        # adata.obs.__setitem__(instance_key, np.arange(adata.n_obs))
     if isinstance(region, str):
-        return TableModel.parse(adata=adata, region=region, instance_key=instance_key)
+        return TableModel.parse(adata=adata, region=region, region_key=region_key, instance_key=instance_key)
```
Comment (giovp) [inline tests/conftest.py:261]: this is also correct and should be readded

File: tests/conftest.py [hunk 4]
```diff
@@ -251,14 +251,16 @@ def _get_table(
     instance_key: Optional[str] = None,
 ) -> AnnData:
     if region is not None:
-        region_key = region_key or "annotated_region"
         instance_key = instance_key or "instance_id"
     adata = AnnData(RNG.normal(size=(100, 10)), obs=pd.DataFrame(RNG.normal(size=(100, 3)), columns=["a", "b", "c"]))
     if instance_key is not None:
+        # from today pycharm refuses to execute this line if I run a test...
```
Comment (giovp) [inline tests/conftest.py:257]: unrelated but there are a bunch of lines of code or comments with # pycharm can or cannot do something around the repo. I would consider to remove them cause they clutter the code and introduce lines/asserts 😅 I would put code quality and consistency with higher priority than whether pycharm works or not in any specific case

**Review by giovp — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 4 above)
Comment (LucaMarconato) [reply tests/conftest.py:257]: yeah I agree; I'll stop using the `mypy` extensions (it's anyway super slow), which is the cause of most of the comments. This one above is actually false, I'll remove it.

File: spatialdata/_core/_spatialdata_ops.py [hunk 5]
```diff
@@ -359,33 +361,34 @@ def _concatenate_tables(tables: list[AnnData]) -> Optional[AnnData]:
...
-            if isinstance(region, list):
-                all_region_keys.append(region_key)
+            if region_key is not None:
+                try:
+                    table.obs[MERGED_TABLES_REGION_KEY] = table.obs[region_key]
+                except KeyError as e:
+                    logger.error(
+                        f"The table has a region_key ({region_key}), but the column with that name is not present in the table"
+                    )
+                    raise e
```
Comment (ivirshup) [inline spatialdata/_core/_spatialdata_ops.py:380]: This should probably be an actual error, not from a logger

**Comment by kevinyamauchi** [comment]
Based on [this comment](https://github.com/scverse/spatialdata/issues/158#issuecomment-1448498092), I think this PR should be updated with the following behavior:

- region_key should always be present, even if region is str
- instance_key type should be consistent across regions. 


I think the filter method can be a follow up PR so that we can merge this quickly and @giovp can get going on the demo datasets.

**Comment by LucaMarconato** [comment]
This PR can be closed without merging as the required changes have been addressed by https://github.com/scverse/spatialdata/pull/164

**Pull request closed without merging**